RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,195 @@
+# Unit checking
+Unit checking is performed by using a type inference algorithm, albeit adapted to work with units. An example of such an algorithm is the Hindley-Milner algorithm.
+At a high level the algorithm collects a set of constraints on unit of expressions. It extracts a sequence of substitutions from these constraints which once applied constructs the unit of a given expression.
+Any inconsistencies discovered during this process, is a unit error, and any unit expression that can't be evaluated is considered undefined.
+
+## Unit of expressions
+
+The unit of an arbitrary Modelica expression is represented by a unit meta-expression.
+This sections outlines the different constructs it comprises.
+
+### Variables
+The unit of variable `var`, is represented by identifier `var.unit`.
+
+### Literals
+
+#### `Well-formed` unit
+A `well-formed` unit can be uniquely represented by its base units factorization, offset and scaling, up to the order of the base units.
+#### `Empty` unit
+An expression with `empty` unit won't contribute to the unit of a parent expression, and will meet any constraints imposed on it.
+#### `Undefined` unit
+An expression with `undefined` unit is used to represent unspecified cases, and will meet any constraints imposed on it.
+
+### Operators
+#### div
+Consider the meta-expression `e1 div e2`.
+* If both `e1` and `e2` are `well-formed`, it evaluates to a `well-formed` unit resulting from dividing `e1` by `e2`.
+* If both `e1` and `e2` are `empty`, it evaluates to `empty`.
+* If `e1` is `empty` and `e2` is well-formed, it evaluates to a `well-formed` unit resulting from dividing `"1"` by `e2`.
+* If `e1` is `well-formed` and `e2` is `empty`, it evaluates to `e1`.
+* If one of them is `undefined`, it evaluates to `undefined`.
+#### mul
+Consider the meta-expression `e1 mul e2`.
+* If both `e1` and `e2` are `well-formed`, it evaluates to a `well-formed` unit resulting from multiplying `e1` by `e2`.
+* If both `e1` and `e2` are `empty`, it evaluates to `empty`.
+* If one of them is `empty`, it evaluates to the other one.
+* If one of them is `undefined`, it evaluates to `undefined`.
+#### pow
+Consider the meta-expression `b pow e`.
+* If `b` is `well-formed` and `e` is an Integer, it evaluates to a `well-formed` unit resulting from raising `b` to the power `e`.
+* If `b` is `"1"`, it evaluates to `"1"`.
+* If `b` is `empty`, it evaluates to `empty`.
+* If `b` is `well-formed` and `e2` is a Real, it evaluates to `undefined`.
+* If `b` is `undefined`, it evaluates to `undefined`.
+
+#### der
+Consider the meta-expression `der(e)`.
+* If `e` is `well-formed`, it evaluates to `e div "s"`.
+* If `e` is `empty`, it evaluates to `empty`.
+* If `e` is `undefined`, it evaluates to `undefined`.
+
+_This definition prevent `der` from propagating `"s"` in a model that does not contain any units. It could lead
+to unwanted unit errors otherwise._
```

Comment:
Review discussions will be easier with sentence-based line breaks:
```suggestion
_This definition prevents `der` from propagating `"s"` in a model that does not contain any units.
It could lead to unwanted unit errors otherwise._
```